### PR TITLE
HPCC-17070 Ensure system can build with USE_ICU=0 and USE_OPENLDAP=0

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately

## Testing:
Checked it compiled.

### CODE REVIEW

**Comment by ghalliday** [comment]
I will run a full regression test, but no code should have been affected if _USE_ICU is defined.  If it isn't defined, then I have implemented a few functions so that utf8->string should work on ascii characters.  I haven't tested it properly though.
@richardkchapman

File: common/thorhelper/thorrparse.cpp [hunk 1]
```diff
@@ -140,6 +142,7 @@ bool isUnicodeMatch(unsigned code, unsigned next)
     case RCCgraph: return u_isprint(next) && !u_isspace(next);
     case RCCpunct: return u_isprint(next) && !(u_isspace(next) || u_isalnum(next));
     case RCCxdigit: return (next < 128) && isxdigit(next);          // should be good enough.
+#endif
     case RCCany:   return true;
     default:
         UNIMPLEMENTED;
```
Comment (richardkchapman) [inline common/thorhelper/thorrparse.cpp]: Should probably be changed to 

      throw MakeStringException(99, "System was built without Unicode support");

File: common/thorhelper/thorrparse.cpp [hunk 2]
```diff
@@ -3063,12 +3073,14 @@ RegexPattern * deserializeRegex(MemoryBuffer & in)
...
         case ThorRegexAsciiSet:     next = new RegexAsciiSetPattern();                  break;
         case ThorRegexAsciiISet:    next = new RegexAsciiISetPattern();                 break;
+#ifdef _USE_ICU
         case ThorRegexUnicode:      next = new RegexUnicodePattern();                   break;
         case ThorRegexUnicodeI:     next = new RegexUnicodeIPattern();                  break;
         case ThorRegexUnicodeSet:   next = new RegexUnicodeSetPattern();                break;
         case ThorRegexUtf8:         next = new RegexUtf8Pattern();                      break;
         case ThorRegexUtf8I:        next = new RegexUtf8IPattern();                     break;
 //      case ThorRegexUnicodeISet:  next = new RegexUnicodeISetPattern();               break;
+#endif
```
Comment (richardkchapman) [inline common/thorhelper/thorrparse.cpp]: default should be changed to

    throw MakeStringException(99, "System was built without Unicode support");

File: ecl/hqlcpp/hqlregex.cpp [hunk 3]
```diff
@@ -878,10 +884,12 @@ bool HqlRegexExpr::canConsume(unsigned nextChar)
                             {
                                 if (options.type == type_unicode)
                                 {
+#ifdef _USE_ICU
                                     //MORE: Improved unicode
                                     if (u_foldCase(nextChar, U_FOLD_CASE_DEFAULT) >= u_foldCase(low, U_FOLD_CASE_DEFAULT) &&
                                         u_foldCase(nextChar, U_FOLD_CASE_DEFAULT) <= u_foldCase(high, U_FOLD_CASE_DEFAULT))
                                         return !invert;
+#endif
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlregex.cpp]: Any need for an else, or is this just an optimization that can be skipped ?

File: ecl/hqlcpp/hqlregex.cpp [hunk 4]
```diff
@@ -1006,6 +1017,9 @@ void HqlRegexExpr::gatherConsumeSymbols(SymbolArray & symbols)
                     UNIMPLEMENTED;
                 }
             }
+#else
+            UNIMPLEMENTED;
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlregex.cpp]: Better error message?

File: ecl/hqlcpp/hqlregex.cpp [hunk 5]
```diff
@@ -1461,7 +1475,11 @@ void HqlRegexExpr::generateRegex(GenerateRegexCtx & ctx)
         {
             RegexSetBasePattern * pattern;
             if (ctx.info.type != type_string)
+#ifdef _USE_ICU
                 pattern = new RegexUnicodeSetPattern(caseSensitive);
+#else
+                UNIMPLEMENTED;
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlregex.cpp]: Better error message

File: ecl/hqlcpp/hqlregex.cpp [hunk 6]
```diff
@@ -1524,6 +1544,8 @@ void HqlRegexExpr::generateRegex(GenerateRegexCtx & ctx)
                         created.setown(new RegexAsciiIPattern(len, data));
                     break;
                 }
+            default:
+                UNIMPLEMENTED;
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlregex.cpp]: Better error message?

File: rtl/eclrtl/eclrtl.cpp [hunk 7]
```diff
@@ -1542,6 +1551,17 @@ void rtlConcatVUnicode(UChar * * tgt, ...)
     buffer[idx++] = 0x0000;
     *tgt = buffer;
 }
+#else
+void rtlConcatUnicode(unsigned & tlen, UChar * * tgt, ...)
+{
+    UNIMPLEMENTED;
```
Comment (richardkchapman) [inline rtl/eclrtl/eclrtl.cpp]: Better error message?

File: rtl/eclrtl/eclrtl.cpp [hunk 8]
```diff
@@ -1542,6 +1551,17 @@ void rtlConcatVUnicode(UChar * * tgt, ...)
...
+}
+
+void rtlConcatVUnicode(UChar * * tgt, ...)
+{
+    UNIMPLEMENTED;
```
Comment (richardkchapman) [inline rtl/eclrtl/eclrtl.cpp]: Better error message

**Review by richardkchapman — CHANGES_REQUESTED** [review]

**Comment by richardkchapman** [comment]
@ghalliday A few comments

**Comment by ghalliday** [comment]
@richardkchapman cleaned up the code and made the error messages consistent.

File: ecl/hqlcpp/hqlregex.cpp (hunk 6 above)
Comment (richardkchapman) [inline ecl/hqlcpp/hqlregex.cpp]: @ghalliday Looks like you missed this one?

Comment (ghalliday) [inline ecl/hqlcpp/hqlregex.cpp]: @richardkchapman I moved the appropriate errors into the case branches - this is UNIMPLEMENTED is correct

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**